### ISSUE TITLE
Resources other than metal/energy

### ISSUE THREAD

**Issue by sprunk** [beyond-all-reason/spring#940]
Collection of misc thoughts that I had written down from some ancient discord discussion. This is all from the **game dev perspective** so doesn't address engine implementation details (including whether this involves ECS or not). 

### Context
 * a game **might simply want to have 3+** resources.
 * a game that only needs 2 is **still made harder to develop** and work with if they are supposed to be gold/wood, or minerals/gas, etc since even if you expose them as such to the player, the **engine interfaces still call them metal/energy**. 
 * there are the vague plans to implement ECS for resources at some point so that may be a good opportunity to touch the wider design. It doesn't rely on it though and both can be done independently.

### How many?
 * of course **ideally there would be no limit**, but there may be perf and design implications.
 * off the top of my head, the **most resources I've seen used in an RTS is 6**. One is the mineralz mod for SC2 which maybe doesn't really count (but that kind of modding is the lifeblood of projects around here!), but also Cossacks was a serious commercial game with 6. HoMM has 7, it's not an RTS, but close and it sounds like the kind of game a crazy modder could want to make an RTS spin-off of. There may also be more resources in other real game examples that I don't personally know. In that vein **I think 8 would be a good limit**, it leaves some room for creativity while being nice and round.
 * the original ECS proposal said 4, which feels a bit harsh and existing-game-centric (in that it would support existing games but not much above), though **if the overhead is significant** it would probably be acceptable. Hard to tell beforehand.
 * perhaps a mixed hard/soft limit? A more lenient hard limit (say 8 or 32 or **whatever is implementationally maximal** without increasing passive overhead) that will probably **allow a game to make tradeoffs** if it wants to, and soft limit that you specify in a modrule (e.g. "this game has 2 resources") that **prevents the engine needlessly processing** the extras.

### Lua API

The basic and obvious approach would be to **make interfaces refer to resources via a numerical index**, grouped into a table where appropriate, with **names read from a modrule and exposed as a `Game` subtable**:
```diff
 modrules.lua
+ resources = { "Gold", "Vespene", "Requisition" }
+ -- Game.Resource.Vespene = 2 and Game.Resource[2] = "Vespene", etc

 unit.lua
- metalCost = 426
- energyCost = 1826381
+ cost = { [Game.Resource.Metal ] =     426, -- perhaps just [Metal] if the fields were exposed as globals
+          [Game.Resource.Energy] = 1826381, }

 unit.lua, different game
+ cost = {
+   [Game.Resource.Food] =  50,
+   [Game.Resource.Wood] =  25,
+   [Game.Resource.Gold] = 175,
+ }

 gadget.lua
- Spring.AddTeamResource(teamID, "m", +200)
+ Spring.AddTeamResource(teamID, Game.Resource.Minerals, +200)

- local bloodUpkeep = UnitDefNames.factory.energyUpkeep -- using energy as blood
+ local bloodUpkeep = UnitDefNames.factory.upkeep[Game.Resource.Blood]
```

### Misc thoughts
 * ideally all resources would be **fungible**. See #939
 * the engine actually assigns colours to resources. This can be seen via `/resbar` and `/sharedialog` where metal and energy are consistently colour-coded as white and yellow respectively. Perhaps further resources can be assigned arbitrary colours though since this sounds quite unimportant.
 * the above two interfaces might also look bad if you have to cram like 8 different sliders/bars into them. Again though, since these are crude "early dev" interfaces perhaps they can just display the first two resources.
 * a game can in theory already do the Lua API changes on its own, under the hard limit of two resources, by applying enough wrapping and field copying. I don't think GF would be too happy if I made ZK a guinea pig this way but a game can be forward-compatible with the suggestion above to some extent (but I expect the design to change with critique, and also such change would make it less compatible with other Spring/Recoil games if done without engine backing).

**Comment by lhog** [comment]
Well this is non-trivial.
a) lots of hardcoded springisms throughout the engine. E.g.: energy as a resource is linked to many in game actions, metal layer is an integral part of every map
b) Not many existing game devs will be happy to update their game defs and Lua wupgetry.

I think the way to implement that is to add ability to have new custom resources, besides M/E, but then the question is how this is superior to a pure Lua and RuleParams implementation.

**Comment by sprunk** [comment]
> a) lots of hardcoded springisms throughout the engine. E.g.: energy as a resource is linked to many in game actions, metal layer is an integral part of every map

Yes, this is #939 but see the list there - ultimately there don't seem to be that many. Note that for example regarding metal layer, there isn't actually a need to add an energy layer, we just need to say it's a general "resource" layer and allow extracting energy (or sucking blood or whatever your resources are) based on that layer. Of course all existing games will only allow extracting metal, and all existing maps happen to place the "resource" on spots that look like metal, but this is fine. At worst one could say that the first resource on the list is "primary" and has some special privileges where it's too hard to generalize, it doesn't have to be crystal perfect to be useful.

> b) Not many existing game devs will be happy to update their game defs and Lua wupgetry.

The proposal doesn't necessarily suggest breaking backward-compatibility in any way. Both keeping compat in engine (just read `metalCost` as a fallback for `cost[1]` specifically, it's a special treatment but it's acceptable) and as a basecontent Lua file that does the necessary work (which then makes the work for game devs boil down to "add this one VFS.Include line in 3 files") sound fairly easy to do.

> I think the way to implement that is to add ability to have new custom resources, besides M/E, but then the question is how this is superior to a pure Lua and RuleParams implementation.

* `AllowUnitBuildStep` is needed for a Lua implementation and measurements say it's quite bad for performance.
* interface consistency within the same game. All resources are accessed the same way, as opposed to 2 resources being handled via engine interfaces and (N-2) via some in-house rules params, with potential small implementation details differences even if you abstract both modes under common functions (for example rules params become public after game over, can we say the same for built-in resources? Note that it is worrisome that the question can even be asked regardless of the actual answer).
* interface consistency across games. Being in the engine enforces a sort of standard that everybody will then adhere to, which makes code reusable. For example recently I added a set of helper functions to check the game mode (`Spring.Utilities.Gametype.is1v1()`, `isFFA()`, etc). BAR adopted it but somebody randomly decided that `is` needs to be capitalized and now these are incompatible (and I couldn't even change ZK to conform because by that time it was already used in maps and I don't have a good way to check maps for this kind of thing). The difference is subtle in that it is literally just uppercase `i` so it may be hard to see why some code is exploding on porting despite both games claiming to have this interface. But I digress; having this sort of code in engine establishes a standard such that you don't have subtle differences that can cause larger problems.
* in general, one of the points of the engine is to provide core RTS functionality such that people don't need to reimplement it from scratch.

> Mindustry has been moving towards more RTS like gameplay and has well over a dozen resources by default

Aren't Mindustry resources actual in-world items though, transported with conveyor belts and whatnot? This is just about the "classic" RTS resources that are magically teleported to and from global storage represented by a number. It doesn't include things like Civ4 resources where having access to resources like Gems or Aluminium is just a boolean state that enables some things in the closest city (but the resources aren't collected to boost a storage number and aren't actually spent on units).

**Comment by PeanutsRevenge** [comment]
> > Aren't Mindustry resources actual in-world items though, transported with conveyor belts and whatnot? This is just about the "classic" RTS resources that are magically teleported to and from global storage represented by a number. It doesn't include things like Civ4 resources where having access to resources like Gems or Aluminium is just a boolean state that enables some things in the closest city (but the resources aren't collected to boost a storage number and aren't actually spent on units).

Yes, that's how Mindustry works, but a Spring mod could do things differently.  Turrets/Units could have a Tungsten variant granting greater HP, copper for higher fire rate (through cooling), titanium for lighter weight and thereby faster movement.
Coal / Oil / Gas / Wood / BioMass could all be used in power plants for energy, which could be used to "refine" uranium for Fission power, or run electrolysis of water to get hydrogen for fusion.

Currently, only Geo's are of value, territory wise for energy, but if there are resources around the map for enhanced energy generation, that changes the landscape of battles.

If the aim is to provide a better platform for the creativity of modders, more flexibility is better, it's just about balancing with the dev load required/engine performance to run.

If it's possible to have no universal limit, allowing each "game" (as in mod) to define it's own limit, this would permit some to have a high limit and trade off performance there, whilst cutting back elsewhere to regain performance or restrict to only high performance systems.

**Comment by Krogoth100** [comment]
This would be a good change. Start with just 2 resources, adjust implementation and make them fungible, so we can try it out thoroughly, then expand.

**Comment by Krogoth100** [comment]
UnitDef's
[Game.Resource.X = y]
form may be not very good for reading, because there is gonna be a lot of it. EnergyUse/Make/Storage etc.
While for cost it makes sense to write it all in a table like:  cost = {metal = 10, wood = 3, snowflakes = 4}
It is not so handy for Use/Make/Storage.
The alternative is for engine to look for <x>Cost/Use/Make/Storage fields where <x> is the name of resource (woodCost = 10). The fields it will be looking will be generated based on resources game rule, low-case. (Could be done via Lua, but if it needs full backward compatibility, this is the way)

Functions like Spring.AddTeamResource() could use a full string name:
AddTeamResource(team, "Wood", 10)
or
AddTeamResource(team, "w", 10)
^Similar to how it works with "m" and "e".
Have all resource names start with different letters => no hash, comparison by first letter, [256] -> resourceIndex map.
Otherwise hash.

**Comment by sprunk** [comment]
> It is not so handy for Use/Make/Storage. The alternative is for engine to look for x_Cost/Use/Make/Storage fields where x is the name of resource (woodCost = 10).

You can keep `woodCost = 10, energyCost = 20` in unit def files and have post-processing turn that into `cost = { wood = 10, energy = 20 }`:

```lua
-- unitdefs_post.lua
for _, unitDef in pairs(UnitDefs) do
    unitDef.cost = unitDef.cost or {}
    unitDef.storage = unitDef.storage or {}
    for resourceName in pairs(Game.Resource) do
        if unitDef[resourceName .. "Cost"] then
            unitDef.cost[resourceName] = unitDef[resourceName .. "Cost"]
        end
        if unitDef[resourceName .. "Storage"] then
            unitDef.storage[resourceName] = unitDef[resourceName .. "Storage"]
        end
    end
end
```

You don't have to write `[Game.Resource.Wood]` every time either, you can turn the contents of `Game.Resource` into globals so you would just write `cost = { [Wood] = 10, [Energy] = 20 }`.

> Functions like Spring.AddTeamResource() could use a full string name
> AddTeamResource(team, "Wood", 10)

Similar story. Put at wupget entry point:
```lua
local originalAddTeamResource = Spring.AddTeamResource
Spring.AddTeamResource = function (teamID, resourceName, amount)
  return originalAddTeamResource(teamID, Game.Resource[resourceName], amount)
end
```

In general, this kind of "dynamic" engine interface makes it harder to write documentation, increases the complexity of engine code, can confuse people and makes games diverge more than they would otherwise. I'm not sure if the convenience of using a string instead of a number/table is worth all that.

**Comment by Krogoth100** [comment]
@sprunk
You _can_, but **should** you? That's the question.
Take simple
cost = { [Wood] = 10, [Energy] = 20 }
A user not too knowledgeable with Lua will write
cost = { Wood = 10, Energy = 20 }
and then wonder why it doesn't work.

I have written lots of engine wrappers myself for my game. This is okay approach, the question is do you want to go for it.
I am in no way affected by implementation of this, because I can always make a wrapper for myself. But from the pov of a newbie this is all different.
**So, are we doing ergonomic interface from the start, or generic for the wrappers?**

I don't think we have consensus on what API should be, because, for one, in GL department I've been asked to make "ideal" interfaces C++ side (exactly the department where wrappers are the most intuitive choice).
I would rather make a lot of wrappers and keep engine code mostly simple, but that is because I am shipping my game with all the Lua stuff that changes engine API, documented. So engine API is one thing, game API is another. A wiki-page for engine API and a wiki-page for game API.

**Comment by marcushutchings** [comment]
Offworld Trading Company is probably a good example of an RTS that has a lot of resources: 13 in this case.

**Comment by Krogoth100** [comment]
I think my game will require 4 resources at least.
Anyone has plans to take on this?

**Comment by marcushutchings** [comment]
There is the possibility that I resume the ECS flow economy work, which would support 4 resource types (thanks to SSE!)

**Comment by marcushutchings** [comment]
It is possible. Though I think the performance cost would be paid in the flexibility of supporting more than 4 rather than whether more thana 4 is used. We would want to keep to a fixed number of resources in engine: say 4 or 8 for example. Whether you use that much or not is another matter. The engine could be rebuilt to support more if it is needed. Essentially we really want to avoid dynamically allocated lists as much as possible. Indirection has significant impact on performance when dealing with scale and it also complicates state transfer (i.e. save/load or if we manage to separate sim/render threads.)

**Comment by Krogoth100** [comment]
Just putting it like a constant or macro would be good enough, imo
